File: leetcode/15_backtracking/112_Path_Sum.go
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func hasPathSum(root *TreeNode, targetSum int) bool {
	if root == nil {
		return false
	}

	return hasPathSumHelper(root, targetSum, 0)
}

func hasPathSumHelper(root *TreeNode, targetSum, currentSum int) bool {
	if root == nil {
		return false
	}
	currentSum += root.Val
	if root.Left == nil && root.Right == nil {
		return currentSum == targetSum
	}
	return hasPathSumHelper(root.Left, targetSum, currentSum) || hasPathSumHelper(root.Right, targetSum, currentSum)
}

File: leetcode/15_backtracking/112_Path_Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root:
            return False
        return self.hasPathSumHelper(root, targetSum, 0)

    def hasPathSumHelper(self, root: TreeNode, targetSum: int, currentSum: int) -> bool:
        if not root:
            return False
        currentSum += root.val
        if not root.left and not root.right:
            return currentSum == targetSum
        return self.hasPathSumHelper(root.left, targetSum, currentSum) or self.hasPathSumHelper(
            root.right, targetSum, currentSum
        )
